library/psa_crypto.c

```diff
@@ -4111,14 +4125,123 @@ static psa_status_t psa_key_derivation_tls12_prf_generate_next_block(
 
     return( status );
 }
+#else
+static psa_status_t psa_key_derivation_tls12_prf_generate_next_block(
+    psa_tls12_prf_key_derivation_t *tls12_prf,
+    psa_algorithm_t alg )
+{
+    psa_algorithm_t hash_alg = PSA_ALG_HKDF_GET_HASH( alg );
+    uint8_t hash_length = PSA_HASH_SIZE( hash_alg );
+    psa_hash_operation_t backup = PSA_HASH_OPERATION_INIT;
+    psa_status_t status, cleanup_status;
+
+    /* We can't be wanting more output after block 0xff, otherwise
+     * the capacity check in psa_key_derivation_output_bytes() would have
+     * prevented this call. It could happen only if the operation
+     * object was corrupted or if this function is called directly
+     * inside the library. */
+    if( tls12_prf->block_number == 0xff )
+        return( PSA_ERROR_CORRUPTION_DETECTED );
+
+    /* We need a new block */
+    ++tls12_prf->block_number;
```

Comment:
How is a leftover `psa_tls12_prf_key_derivation_t` structure handled after a failure? Since the block number is incremented first, I imagine that it has to be discarded?

Reply:
This is an internal function and we abort the operation on failure up in the call stack before the struct reaches the user. Keeping a consistent state would be indeed more robust, but needs extra checks and operations. The overhead is not big in this instance, but it adds up and can mean a considerable difference in code size if we want to do it everywhere.

Reply:
I'm fine with that, I just wanted to make sure that it's a conscious decision that the struct cannot be reused.